File: leetcode_2708.java
/*
URL = https://leetcode.com/problems/maximum-strength-of-a-group/
2708. Maximum Strength of a Group

Multiply all the positive values anyways
Multiply all the negative values evenly ... with the greatest beg values first

Hang on ... we can get case of all values equal to 0
Be very careful
Forming a group : can we avoid choosing any numbers too ... no ... must be a nonEMpty group
Seems negative positive split fares better too?
And the zero split too?
*/
class Solution {
    public long maxStrength(int[] nums) {
        long myMaxStrength = 1;
        Arrays.sort(nums);
        List<Integer> negNums = new ArrayList<Integer>();
        List<Integer> zeros = new ArrayList<Integer>();
        List<Integer> posNums = new ArrayList<Integer>();
        for(int x : nums) {
            if(x<0){
                negNums.add(x);
            } else if ( x == 0 ) {
                zeros.add(x);
            } else {
                posNums.add(x);
            }
        }
        if(negNums.size() == 0 && posNums.size() == 0) {
            return 0;
        } else {
            if(posNums.size() == 0 && negNums.size() == 1) {
                if(zeros.size() > 0){
                    return 0;
                }
                myMaxStrength = (long)(negNums.get(0));
                // no breka in switch or loop statements too!
            } else {
                for(int x : posNums) {
                    myMaxStrength *= x;
                }
                int i = 0;
                while(i < negNums.size()){
                    int nextIdx = i + 1;
                    if(nextIdx >= negNums.size()){
                        break;
                    }
                    myMaxStrength *= (negNums.get(i) * negNums.get(nextIdx));
                    i += 2;
                }
            }
        }
        return myMaxStrength;
    }
}
Console
/*
URL = https://leetcode.com/problems/maximum-strength-of-a-group/
2708. Maximum Strength of a Group

Multiply all the positive values anyways
Multiply all the negative values evenly ... with the greatest beg values first

Hang on ... we can get case of all values equal to 0
Be very careful
Forming a group : can we avoid choosing any numbers too ... no ... must be a nonEMpty group
Seems negative positive split fares better too?
And the zero split too?
*/
class Solution {
    public long maxStrength(int[] nums) {
        long myMaxStrength = 1;
        Arrays.sort(nums);
        List<Integer> negNums = new ArrayList<Integer>();
        List<Integer> zeros = new ArrayList<Integer>();
        List<Integer> posNums = new ArrayList<Integer>();
        for(int x : nums) {
            if(x<0){
                negNums.add(x);
            } else if ( x == 0 ) {
                zeros.add(x);
            } else {
                posNums.add(x);
            }
        }
        if(negNums.size() == 0 && posNums.size() == 0) {
            return 0;
        } else {
            if(posNums.size() == 0 && negNums.size() == 1) {
                if(zeros.size() > 0){
                    return 0;
                }
                myMaxStrength = (long)(negNums.get(0));
                // no breka in switch or loop statements too!
            } else {
                for(int x : posNums) {
                    myMaxStrength *= x;
                }
                int i = 0;
                while(i < negNums.size()){
                    int nextIdx = i + 1;
                    if(nextIdx >= negNums.size()){
                        break;
                    }
                    myMaxStrength *= (negNums.get(i) * negNums.get(nextIdx));
                    i += 2;
                }
            }
        }
        return myMaxStrength;
    }
}

File: leetcode_2708.go
/*
URL := https://leetcode.com/problems/maximum-strength-of-a-group/description/
2708. Maximum Strength of a Group

Can we avoid two seperate arrays

Complexity
Let N := len(nums)
Time := O(Nlgn)
Space := O(1) ( EXP & IMP ) 

Handle the zero number case too ( can arise ) 

Can have case of all numbers being negative too ( take note of this ) 
Try seperating out at least
strength is additive -> NOP on the zero still

Must be a non-empty group of students too

Cases
(A) [-1] => -1
(B) [-3,-2] => 6
(C) [-3,-2,-1] => 6
(D) [-1,0] => 0
(E) [0] => 0
(F) [1] => 1
(G) [0,0,0,0] => 0
(H) [-3,-2,0,0,1,2] => 6*1*2 = 12
(I)
(J) 
*/
func maxStrength(nums []int) int64 {
    // BASE CASE HANDLING
    if(len(nums) == 1) {
        return (int64)(nums[0])
    }
    maxStren := int64(1)
    n := len(nums)
    ptr := 0
    sort.Ints(nums) // ASC order
    hitTwoNeg := false
    hitPos := false
    for ptr < n {
        if nums[ptr] < 0 {
            next := ptr + 1
            if isInBounds(next,n) && nums[next] < 0 {
                hitTwoNeg = true
                maxStren *= (int64)(nums[ptr] * nums[next])
                ptr += 2
            } else {
                ptr++
            }
        } else if nums[ptr] > 0 {
            hitPos = true
            maxStren *= (int64)(nums[ptr])
            ptr++
        } else if nums[ptr] == 0 {
            ptr++ // NOP here
        }
    }
    // only hit zeroes in this case
    if !hitTwoNeg && !hitPos {
        maxStren = int64(0)
    }
    return maxStren
}

func isInBounds(a int, n int) bool {
    return (0 <= a && a < n)
}
